src/main/scala/monix/nio/tcp/AsyncServerSocketChannel.scala

```diff
@@ -0,0 +1,173 @@
+package monix.nio.tcp
+
+import java.net.{ InetSocketAddress, StandardSocketOptions }
+import java.nio.channels.spi.AsynchronousChannelProvider
+import java.nio.channels.{ AsynchronousChannelGroup, AsynchronousServerSocketChannel, AsynchronousSocketChannel, CompletionHandler }
+
+import monix.eval.{ Callback, Task }
+import monix.execution.Scheduler
+import monix.nio.internal.ExecutorServiceWrapper
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+/**
+ * An asynchronous channel for stream-oriented listening sockets.
+ *
+ * On the JVM this is a wrapper around
+ * [[https://docs.oracle.com/javase/8/docs/api/java/nio/channels/AsynchronousServerSocketChannel.html java.nio.channels.AsynchronousServerSocketChannel]]
+ * (class available since Java 7 for doing async I/O on sockets).
+ *
+ * @example {{{
+ *   val server = AsyncServerSocketChannel()
+ *   server.bind(new InetSocketAddress(InetAddress.getByName(null), 9000))
+ *
+ *   val bytes = ByteBuffer.wrap("Hello world!".getBytes("UTF-8"))
+ *   val writeF = server
+ *     .accept()
+ *     .flatMap { conn =>
+ *       val writeF0 = conn.write(bytes, None)
+ *       conn.stopWriting()
+ *       writeF0
+ *     }
+ *     .map { sentLen =>
+ *        server.close()
+ *        sentLen
+ *     }
+ *
+ *   writeF.onComplete {
+ *     case Success(nr) =>
+ *       println(f"Bytes sent: $nr%d")
+ *
+ *     case Failure(exc) =>
+ *       println(s"ERR: $exc")
+ *   }
+ * }}}
+ *
+ * @define callbackDesc is the callback to be called with the result, once
+ *         this asynchronous operation is complete
+ *
+ * @define acceptDesc Accepts a connection
+ *
+ * @define bindDesc Binds the channel's socket to a local address and configures the socket to listen for connections
+ * @define localDesc the local address to bind the socket, or null to bind to an automatically assigned socket address
+ * @define backlogDesc the maximum number of pending connections. If the backlog parameter has the value 0,
+ *         or a negative value, then an implementation specific default is used.
+ *
+ * @define localAddressDesc Asks the socket address that this channel's socket is bound to
+ */
+abstract class AsyncServerSocketChannel extends AutoCloseable {
+
+  /**
+   * $acceptDesc
+   *
+   * @param cb $callbackDesc
+   */
+  def accept(cb: Callback[AsyncSocketChannel]): Unit
+
+  /**
+   * $acceptDesc
+   */
+  def accept(): Future[AsyncSocketChannel] = {
+    val p = Promise[AsyncSocketChannel]()
+    accept(Callback.fromPromise(p))
+    p.future
+  }
+
+  /**
+   * $acceptDesc
+   */
+  def acceptL(): Task[AsyncSocketChannel] =
```

Comment:
This is based on `accept(Callback)`. Maybe it would be good to separate the types, such that we have `AsyncServerSocketChannel` (dealing with `Future` along with low level callbacks) and `TaskServerSocketChannel`, dealing with `Task`.

We've made this separation in Monix as well, for example in `AsyncSemaphore` (in `monix-execution`) versus `TaskSemaphore` (in `monix-eval`).